libraries/chez/common/immutable/list.effekt

```diff
@@ -3,124 +3,599 @@ module immutable/list
+    list match {
+      case Nil() => ();
+      case Cons(head, tail) =>
+        f(i, head)
+        go(tail, i+1){f}
+    }
   }
 
-  loop(l, 0)
+  go(list, 0){f}
+}
+
+// Map a function `f` over elements in a given list.
+//
+// O(N)
+def map[A, B](l: List[A]) { f: A => B } : List[B] = {
+  var acc = Nil[B]()
+  l.foreach { el => acc = Cons(f(el), acc) }
+  acc.reverse
 }
 
+// Map a function `f` over elements in a given list,
+// keeping only the elements for which the function returned `Some(...)`,
+// discarding the elements for which the function returned `None()`.
+//
+// O(N)
 def collect[A, B](l: List[A]) { f : A => Option[B] }: List[B] = {
-  def loop(lst: List[A], acc: List[B]): List[B] = lst match {
-    case Nil() => acc
-    case Cons(a, as) =>
-      val optB = f(a)
-      optB match {
-        case None() => loop(as, acc)
-        case Some(b) => loop(as, Cons(b, acc))
-      }
+  var acc = Nil[B]()
+  l.foreach { a =>
+    val optB = f(a)
+    optB match {
+      case None() => ();
+      case Some(b) => acc = Cons(b, acc);
+    }
   }
-  loop(l, Nil()).reverse
+  acc.reverse
 }
 
+// Map a function `f` over elements in a given list and concatenate the results.
+//
+// O(N)
 def flatMap[A, B](l: List[A]) { f : A => List[B] }: List[B] = {
-  def loop(lst: List[A], acc: List[B]): List[B] = lst match {
-    case Nil() => acc
-    case Cons(a, as) =>
-      val bs = f(a)
-      loop(as, acc.append(bs))
+  var acc = Nil[B]()
+  l.foreach { a =>
+    val bs = f(a)
+    acc = acc.append(bs)
   }
-  loop(l, Nil())
+  acc
 }
 
-def sum(l: List[Int]): Int =
-  l match {
-    case Nil() => 0
-    case Cons(a, as) => a + sum(as)
+// Check if predicate is true for all elements of the given list.
+//
+// O(N)
+def all[A](list: List[A]) { predicate: A => Boolean }: Boolean = {
+  list match {
+    case Cons(x, xs) => predicate(x) && all(xs) { predicate }
+    case Nil() => true
   }
+}
+
+// Check if predicate is true for at least one element of the given list.
+//
+// O(N)
+def any[A](list: List[A]) { predicate: A => Boolean }: Boolean = {
+  list match {
+    case Cons(x, xs) => predicate(x) || any(xs) { predicate }
+    case Nil() => false
+  }
+}
 
-def size[A](l: List[A]): Int = l match {
-  case Nil() => 0
-  case Cons(_, rest) => 1 + rest.size
+// Sum the elements of the list.
+//
+// O(N)
+def sum(list: List[Int]): Int = {
+  var n = 0;
+  list.foreach { x => n = n + x };
+  n
 }
 
+// Calculate the size of the list.
+//
+// O(N)
+def size[A](l: List[A]): Int = {
+  var n = 0;
+  l.foreach { _ => n = n + 1 };
+  n
+}
+
+// Reverse the list.
+//
+// O(N)
 def reverse[A](l: List[A]): List[A] = {
-  def reverseWith(l: List[A], acc: List[A]): List[A] = l match {
-    case Nil() => acc
-    case Cons(a, rest) => rest.reverseWith(Cons(a, acc))
-  }
-  l.reverseWith(Nil())
+  var res = Nil[A]()
+  l.foreach { el => res = Cons(el, res) }
+  res
 }
 
-// [1,2,3].reverseOnto([4,5,6])
+// Reverse a list `l` and append it to `other`.
```

Comment:
Wording: `append` should be `prepend`

Reply:
Thanks! :)

Reply:
I decided to use:
```
// Reverse a list `l` and append `other` to it.
```
to avoid using the word `prepend` which is not used anywhere else. :)

However, let me know if you have a better description.
To be honest, I'm not sure that a function like this needs to be in the stdlib... 🤔 (it's only used in the definition of append, IIRC)

Reply:
We might also want to consider a similar approach with swapped arguments used in the [standard library of Idris 2](https://github.com/idris-lang/Idris2/blob/0e831ed5ef6670d01e025197f9d1b3178586be9b/libs/base/Data/List1.idr#L113)